plugins/spidmx/SpiDmxParser.cpp

```diff
@@ -0,0 +1,538 @@
+/*
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Library General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * SpiDmxParser.cpp
+ * This parses a SPI buffer into a DmxBuffer and notifies a callback when a
+ * packet is received completely.
+ * Copyright (C) 2017 Florian Edelmann
+ */
+
+#include <stdio.h>
+
+#include "ola/Callback.h"
+#include "ola/DmxBuffer.h"
+#include "ola/Logging.h"
+#include "plugins/spidmx/SpiDmxParser.h"
+
+namespace ola {
+namespace plugin {
+namespace spidmx {
+
+/**
+ * This class implements the DMX protocol on a very low level, so be sure to
+ * fully understand the protocol before you tackle this code ;)
+ *
+ * Possible DMX frequencies are 245 - 255kbit/s.
+ * With a sampling rate of 2MHz, this results in one DMX bit being
+ * mapped to 8.163265306122449 - 7.843137254901961 SPI bits.
+ * So if we calculate with 7.5 - 8.5 bit length, we should be fine.
+ *
+ * Abbreviations used in this class:
+ *  - MAB: Mark after break
+ *  - MBS: Mark between slots
+ *  - MBB: Mark before break
+ */
+
+
+/**
+ * Loop through the given SPI raw bytes and call in every iteration the
+ * respective function that handles the byte in the current state.
+ *
+ * @param *buffer - The buffer with SPI bytes to read from
+ * @param buffersize - Size of the buffer
+ */
+void SpiDmxParser::ParseDmx(uint8_t *buffer, uint64_t buffersize) {
+  chunk = buffer;
+
+  chunk_bitcount = 0;
+
+  ChangeState(WAIT_FOR_BREAK);
+
+  while (chunk_bitcount < buffersize) {
+    switch (state) {
+      case WAIT_FOR_BREAK:
+        // printf("%6ld  0x%02x  wait for break (%d)\n", chunk_bitcount,
```

Comment:
Can we remove this debug please, or convert it to OLA_DEBUG.

Reply:
If I put it in OLA_DEBUG, it outputs so much that receiving gets too laggy to be usable. I'll remove those comments shortly before merging but I keep them until then for easy debugging.